### PR TITLE
[ADD] core: fields computed before insertion in database

### PR DESCRIPTION
**Pull request by rco-odoo**
Given value | Field.default | Field.compute | Current effect | Expected effect
--- | --- | --- | --- | ---
none | none | none | none
value | none | none | INSERT value
none | default | none | INSERT default
value | default | none | INSERT value
none | none | compute readonly | no INSERT, post-compute | INSERT computed value
value | none | compute readonly | no INSERT, post-compute | INSERT computed value
none | default | compute readonly | no INSERT, post-compute | INSERT computed value
value | default | compute readonly | no INSERT, post-compute | INSERT computed value
none | none | compute read-write | no INSERT, post-compute | INSERT computed value
value | none | compute read-write | INSERT value
none | default | compute read-write | INSERT default
value | default | compute read-write | INSERT value
none | none | compute with inverse | no INSERT, post-compute | INSERT computed value
value | none | compute with inverse | INSERT value, then INVERSE field
none | default | compute with inverse | INSERT default, then INVERSE field
value | default | compute with inverse | INSERT value, then INVERSE field

### CODE REVIEW

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 1]
```diff
@@ -3411,3 +3431,112 @@ class Foo(models.Model):
...
+        with self.assertQueries(QUERIES):
+            record = model.create({'name': 'A'})
+
+        self.assertEqual(record.reader, 'A')
+        self.assertEqual(record.editer, 'A')
+        self.assertEqual(record.setter, 'A')
+
+        # default value
+        with self.assertQueries(QUERIES), self.assertLogs('precompute_setter', level='WARNING'):
+            record = model.with_context(reader='X', editer='Y', setter='Z').create({'name': 'A'})
```
Comment (Feyensv) [inline odoo/addons/test_new_api/tests/test_new_fields.py:3530]:
```suggestion
            record = model.with_context(default_reader='X', default_editer='Y', default_setter='Z').create({'name': 'A'})
```
?

**Comment by ryv-odoo** [comment]
Nice job! It can improve the performance of some flows.

It is legit to keep the `precompute` to `False` as default (even for related store field) because, there are some tricky counterproductive cases, by example:
On a Model A, with a computed field X depending on a Many2one field Y_id (all related ?). With some flow where the `create` isn't batched (shame) it can lead to extra SQL request to fetch data of `Y` model:
```python
@api.depends('Y.name', 'Y.code')
def _compute_X(self):
    for a in self:
        a.X = f"{Y.name} {Y.code}"
...

# with a no-batched flow:
    ...
    for y in Y.search(...):
        # In this case, the precompute will fetch data (prefeched fields) of Y record by record (one SQL request by `y`) 
        a = A.create({'Y_id': y.id})  
        ...
```

**Comment by Feyensv** [comment]
> It is legit to keep the precompute to False as default (even for related store field) because, there are some tricky counterproductive cases, by example:

The overhead of the new record emulation may not be light depending on model complexity and given values.
We decided this change would be for a future task, after the conversion of multiple core & heavy models, and extensive testing.

**Comment by tivisse** [comment]
Note: Pull request linked to 
https://github.com/odoo/odoo/pull/79093
https://github.com/odoo/enterprise/pull/21930

**Comment by rco-odoo** [comment]
@robodoo r+

**Pull request closed without merging**